src/main/java/org/hashids/Hashids.java

```diff
@@ -23,154 +29,164 @@
 
-    String guards;
-    if (alphabet.length() < 3) {
-      guards = seps.substring(0, guardCount);
-      seps = seps.substring(guardCount);
+    char[] guards;
+    if (alphabet.length < 3) {
+      guards = copyOf(seps, guardCount);
+      seps = copyOfRange(seps, guardCount, seps.length);
     } else {
-      guards = alphabet.substring(0, guardCount);
-      alphabet = alphabet.substring(guardCount);
+      guards = copyOf(alphabet, guardCount);
+      alphabet = copyOfRange(alphabet, guardCount, alphabet.length);
     }
+
     this.guards = guards;
     this.alphabet = alphabet;
     this.seps = seps;
+    this.validChars = concatenate(alphabet, guards, seps);
+    this.guardsRegExp = '[' + String.valueOf(guards) + ']';
+    this.sepsRegExp = '[' + String.valueOf(seps) + ']';
+  }
+
+  private void validateAlphabet(char[] alphabet) {
+    if (alphabet.length < MIN_ALPHABET_LENGTH) {
+      throw new IllegalArgumentException(
+          "The alphabet must contain at least " + MIN_ALPHABET_LENGTH + " unique characters.");
+    }
+
+    for (int i = 0; i < alphabet.length; i++) {
+      if (isSpaceChar(alphabet[i])) {
+        throw new IllegalArgumentException("The alphabet cannot contain spaces.");
+      }
+
+      if ((i + 1 < alphabet.length) && (alphabet[i] == alphabet[i + 1])) {
+        throw new IllegalArgumentException("The alphabet cannot contain duplicates.");
+      }
+    }
   }
 
+  // ---------
+
   /**
    * Encode numbers to string
    *
-   * @param numbers
-   *          the numbers to encode
+   * @param numbers the numbers to encode
    * @return the encoded string
    */
   public String encode(long... numbers) {
     if (numbers.length == 0) {
-      return "";
+      throw new IllegalArgumentException("At least one number must be specified.");
     }
 
     for (final long number : numbers) {
       if (number < 0) {
-        return "";
+        return ""; // we must throw an exception here (like the case when we compare with MAX_NUMBER)
       }
+
       if (number > MAX_NUMBER) {
-        throw new IllegalArgumentException("number can not be greater than " + MAX_NUMBER + "L");
+        throw new IllegalArgumentException("Number can not be greater than " + MAX_NUMBER + '.');
       }
     }
+
     return this._encode(numbers);
   }
 
   /**
    * Decode string to numbers
    *
-   * @param hash
-   *          the encoded string
+   * @param hash the encoded string
    * @return decoded numbers
    */
   public long[] decode(String hash) {
     if (hash.isEmpty()) {
       return new long[0];
     }
-    
-    String validChars = this.alphabet + this.guards + this.seps;
-    for (int i = 0; i < hash.length(); i++) {
-      if(validChars.indexOf(hash.charAt(i)) == -1) {
-        return new long[0];
-      }
+
+    if (!validate(hash.toCharArray(), validChars)) {
+      return new long[0];
     }
 
-    return this._decode(hash, this.alphabet);
+    return _decode(hash, alphabet);
```

Comment:
always prefix methods/attributes with `this.`

Reply:
Why? This is not a JAVA rule. Perhaps a JS one.